### DISCUSSION TITLE
Validate body request field with database query

### DISCUSSION THREAD

**Discussion by bert2002** [fastapi/fastapi#3924] [Questions] [answered]
Hi,
is it possible to verify a field in the body request by running a sql query? E.g.:

```
@app.post("/{x}")
def get_model(
  source: ModelSource,
  data: SourceData,
  user: User = Depends(current_active_verified_user)
  ):
```
and my source data looks like this:
```
class SourceData(BaseModel):
  field1: int
  field2: str
```
Now what I would like to do is to check if `field1` is a int that I can verify in my database.

Any idea or hints how this could be done?

Thanks,
bert

**Comment by ghandic** [comment]
This is possible using pydantic validators, note the check uses logic to check if it is in a list but you could do any logical check there you need

```python
from fastapi import FastAPI, APIRouter, Depends, HTTPException
from pydantic import BaseModel, validator

app = FastAPI(docs_url="/")
router = APIRouter()

db = ["user_001"]


class CreateUser(BaseModel):
    username: str
    password: str

    @validator("username")
    def username_in_db(cls, v):
        if v not in db:
            raise HTTPException(status_code=403, detail="Username not found")
        return v


@router.post("/login")
def login(login: CreateUser = Depends(CreateUser)):
    return login


app.include_router(router=router, prefix="/example")


```

**Comment by YuriiMotov** [comment, marked as answer]
There is no standard way to achieve this in FastAPI.
Common approach is to implement this validation in endpoint function or in dependency and raise error status code `400`

**Reply by ghandic** [reply]
Isn't that what is in more detail above?

**Reply by YuriiMotov** [reply]
As I see it, it's not.
It was asked how to use results of DB query to validate model. To achieve this using `validator` you would need to somehow pass db session to it (or the result of other dependency), that is impossible for now (I assume we are not going to teach people bad practices of using global variable fir this reason).

**Reply by ghandic** [reply]
Seemed to satisfy the OP 4 years ago based on the thumbs up, imo your original point added no more value and you marked it as the answer?

**Reply by YuriiMotov** [reply]
> imo your original point added no more value

It's you opinion. I think it's wrong.
